syntax = "proto3";

package client_server;

service ChatApp {
   	rpc EnterChatRoom(stream Message) returns (stream Message) {}

	rpc SendRequest(Request) returns (RequestResult) {}

	rpc VoteRequest(VotingInfo) returns (Vote) {}
}


message User {
	string userID = 1;
}

message Message {
	string content = 1;
	User sender = 2;
	string time = 3;
}

message ChatRoom {
    string chatRoomID = 1;
}

message Request {
    
    enum RequestType {
      ENTER = 0;
      EXIT = 1;
      CHANGE_NAME = 2;
    }
    
    RequestType type = 1;
    ChatRoom chatRoom = 2;
    string info = 3; // i.e. new nickname
}

message RequestResult {
    bool success = 1;
    Request request = 2;
}

message VotingInfo {
    User requester = 1;
    ChatRoom chatRoom = 2;
}

message Vote {
    bool approved = 1;
    VotingInfo voteInfo = 2;
    User voter = 3;
}
